commit f9755e392eed49c4b230b32704bcc554950114c3
Author: Pranav Vashi <neobuddy89@gmail.com>
Date:   Mon Sep 29 02:19:17 2025 +0530

    ActivityManagerService: Perform system compaction with GC
    
    * Do not run this on screen-on path - as it is bit CPU intensive.
    
    Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
    Signed-off-by: Ghosuto <clash.raja10@gmail.com>

diff --git a/core/java/android/app/IActivityManager.aidl b/core/java/android/app/IActivityManager.aidl
index f7e285127656..c96d31dbde40 100644
--- a/core/java/android/app/IActivityManager.aidl
+++ b/core/java/android/app/IActivityManager.aidl
@@ -1068,4 +1068,6 @@ interface IActivityManager {
     boolean shouldForceCutoutFullscreen(in String packageName);
 
     void releaseMemory(int minAdj, int maxKillCount, boolean includeUIProcesses, boolean skipCamera);
+
+    void compactAllSystem();
 }
diff --git a/services/core/java/com/android/server/am/ActivityManagerService.java b/services/core/java/com/android/server/am/ActivityManagerService.java
index dd6912a68479..ebc460a17a9c 100644
--- a/services/core/java/com/android/server/am/ActivityManagerService.java
+++ b/services/core/java/com/android/server/am/ActivityManagerService.java
@@ -19960,8 +19960,6 @@ public class ActivityManagerService extends IActivityManager.Stub
 
         synchronized (this) {
             synchronized (mProcLock) {
-                mCachedAppOptimizer.compactAllSystem();
-
                 mProcessList.forEachLruProcessesLOSP(false, proc -> {
                     if (proc == null || proc.getThread() == null) return;
 
@@ -19997,6 +19995,15 @@ public class ActivityManagerService extends IActivityManager.Stub
         }
     }
 
+    @Override
+    public void compactAllSystem() {
+        mHandler.post(() -> {
+            synchronized (mProcLock) {
+                mCachedAppOptimizer.compactAllSystem();
+            }
+        });
+    }
+
     public class ProcessComparator implements Comparator<ProcessToKill> {
         @Override
         public int compare(ProcessToKill p1, ProcessToKill p2) {
diff --git a/services/core/java/com/android/server/policy/PhoneWindowManager.java b/services/core/java/com/android/server/policy/PhoneWindowManager.java
index 0904455dd294..aecd4383778e 100644
--- a/services/core/java/com/android/server/policy/PhoneWindowManager.java
+++ b/services/core/java/com/android/server/policy/PhoneWindowManager.java
@@ -433,7 +433,7 @@ public class PhoneWindowManager implements WindowManagerPolicy {
     private static final long MEMORY_RELEASE_INTERVAL_MS = 5 * 60 * 1000L; // 5 minutes
     private long lastMemoryReleaseTime = 0L;
 
-    private static final long GC_INTERVAL_MS = 10 * 60 * 1000L; // 10 minutes
+    private static final long GC_INTERVAL_MS = 5 * 60 * 1000L; // 5 minutes
     private long lastGcTime = 0L;
 
     /**
@@ -6510,8 +6510,12 @@ public class PhoneWindowManager implements WindowManagerPolicy {
                 System.gc();
                 System.runFinalization();
                 System.gc();
+                try {
+                    mActivityManagerService.compactAllSystem();
+                } catch (RemoteException e) {
+                }
                 lastGcTime = currentTime;
-                Log.v("GcOpt", "performing garbage collection for system_server");
+                Slog.d(TAG, "Performing garbage collection for system_server");
             }
         }
     };
